Add CI for a Python wheel release to PyPi

mballance opened the pull request:
Signed-off-by: Matthew Ballance <matt.ballance@gmail.com>

mballance commented:
I've updated the Azure Pipelines specification to only authenticate and upload to PyPi when the build is done from the 'master' branch. It's pretty unlikely that we want PR builds to make uploads to PyPi.

In order to enable automated uploads to pypi.org, the following must be done in the Azure Pipelines Builds section. Before doing so, register for a free account on pypi.org
- Project Settings -> Service Connections (it's under the Pipelines heading)
- New Service connection -> Python package upload
  - Connection name: pypi
  - Python repository url for upload: https://pypi.python.org/pypi
  - EndpointName: pyboolector
  - Username: pypi.org username
  - Password: pypi.org password
- Select OK

The CI script can now authenticate a connection to PyPi.org via the information saved in the service connection configuration.

mpreiner commented:
@mballance Thanks a lot for the work and sorry for the delay! We'll have a look and setup pypi ASAP.

mpreiner commented:
@mballance Sorry that this takes so long. I had a brief look (will dive into it more soon). However, I have one question: Why did you choose to use Apache instead of sticking with MIT that comes with Boolector? This may cause some confusion since only the code for packaging pyboolector for pypi would be under Apache, but Boolector and pyboolector are still MIT, which is currently not reflected in setup.py.

mballance commented:
@mpreiner, looks like an oversight on my part. I'll be happy to correct the license to align with the MIT that Boolector uses.

-Matthew

mpreiner commented:
Thanks @mballance!

mpreiner commented on pypi/LICENSE:
```diff
@@ -0,0 +1,27 @@
+Boolector is a Satisfiability Modulo Theories (SMT) Solver for bit-vectors,
```
Can we use https://github.com/Boolector/boolector/blob/master/COPYING instead of having a separate copy of the license?

mballance replied:
Good point. I'll update the build script to grab COPYING from the root of the tree

mpreiner commented on pypi/build.sh:
```diff
@@ -0,0 +1,173 @@
+echo "Hello from PyPi build.sh"
+
+BUILD_DIR=`pwd`
+if test -f /proc/cpuinfo; then
+  echo "cpuinfo exists"
+  N_CORES=`cat /proc/cpuinfo | grep processor | wc -l`
+else
+  echo "cpuinfo DOES NOT exist"
+  N_CORES=1
+fi
```
```suggestion
N_CORES=`nproc`
```

mpreiner commented on pypi/build.sh:
```diff
@@ -0,0 +1,173 @@
+
+cp -r /boolector .
+
+if test ! -f lingeling-master.zip; then
+  curl -L -o lingeling-master.zip https://github.com/arminbiere/lingeling/archive/master.zip
+  if test $? -ne 0; then exit 1; fi
+fi
```
Was there a specific reason why you didn't use the scripts in `contrib/` for setting up the dependencies? The issue with this is that you are always downloading master, but the `contrib` scripts define specific solver versions in order to not break the builds in case something changes upstream in the dependencies.

mballance replied:
Yes. The commit of lingeling that Boolector uses causes a compile error with the GCC used for Python. This has been fixed on 'master'. Perhaps we could update to using a newer commit?

mballance replied:
BTW, I've updated the lingeling commit in the PR

mpreiner commented on pypi/build.sh:
```diff
@@ -0,0 +1,173 @@
+  if test $? -ne 0; then exit 1; fi
+fi
+
+if test ! -f btor2tools-master.zip; then
```
Ditto.

mpreiner commented on pypi/build.sh:
```diff
@@ -0,0 +1,173 @@
+if test $? -ne 0; then exit 1; fi
+cp src/btor2parser/btor2parser.h  ${BUILD_DIR}/boolector/deps/install/include/btor2parser
+if test $? -ne 0; then exit 1; fi
+
```
It would be great if you also include CaDiCaL for the best possible performance.

mballance replied:
Yes, will do.

mpreiner commented on pypi/build.sh:
```diff
@@ -0,0 +1,173 @@
+#********************************************************************
+cd ${BUILD_DIR}
+
+sed -i -e 's/add_check_c_cxx_flag("-W")/add_check_c_cxx_flag("-W")\nadd_check_c_cxx_flag("-fPIC")/g' \
+  boolector/CMakeLists.txt
```
Why is this necessary? `./configure.sh --python --shared` should take care of this?

mballance replied:
The OS used by the docker image used to build PyPi packages has Python 2.7 as default, and this creates issues with Boolector's built-in Python-extension build. Looks like simply passing --shared works okay, though

mpreiner commented on pypi/build.sh:
```diff
@@ -0,0 +1,173 @@
+
+cd boolector
+
+./configure.sh -fPIC --shared --prefix /usr
```
`-fPIC` shouldn't be needed here since you already pass `--shared`.

mballance replied:
Good point. Looks like --shared is sufficient

mpreiner commented on pypi/build.sh:
```diff
@@ -0,0 +1,173 @@
+#!/bin/sh
+
```
```suggestion
set -e -o pipefail
```
Then you don't need the `if test $? -ne 0; then exit 1; fi` checks anymore.

mballance replied:
Good point. Change made

mpreiner reviewed (changes requested):
@mballance as discussed offline here is the first pass over the PR. It seems that the build script can be simplified quite a bit and I'd like to include CaDiCaL, which is the default SAT solver for Boolector and provides better performance. Thanks again for your work and sorry for the delay!

mballance commented:
Hi @mpreiner,
    I've updated the PR with your requested changes, and also added in a couple of changes from the separate CI build -- macOS release, and improved versioning.
- The PR now uses the contrib/setup scripts for building dependencies. I updated the lingeling commit hash to pick up the fix for a name-collision bug (https://github.com/arminbiere/lingeling/pull/2) that impacted the PyPi builds
- Both CaDiCaL and Lingeling SAT solvers are provided
- Simplified the build script for Linux, and made a few updates to Docker image and CI config

Thanks and Best Regards,
Matthew

andrewvaughanj commented:
My Lingeling PR isn’t merged; how are you using the commit hash from that?

Could we do something here like we do with the Windows patches to make this self-contained within the Boolector repo?

andrewvaughanj commented on pypi/build.sh:
```diff
@@ -0,0 +1,79 @@
+cd boolector
+
+./configure.sh --shared --prefix /usr
+if test $? -ne 0; then exit 1; fi
```
Now you’ve got `set -e` at the top you shouldn’t need this check.

mballance replied:
Yup, looks like I missed removing a couple. Hopefully all clean now

mballance commented:
> My Lingeling PR isn’t merged; how are you using the commit hash from that?
> 
> Could we do something here like we do with the Windows patches to make this self-contained within the Boolector repo?

Apologies for the confusion, @andrewvaughanj. I wanted to point @mpreiner to a description of the issue resolved by moving to the latest lingeling code. It appears the issue has been resolved independent of your PR. The clone variable is now named 'cloned'.

andrewvaughanj commented on src/api/python/pyboolector.pyx:
```diff
@@ -6,6 +6,7 @@
 # This file is part of Boolector.
 # See COPYING for more information on using this software.
 #
+# cython: language_level=3
```
This will affect Python 2 users (who shouldn’t exist, but they do).

Can’t we just invoke Cython with the correct language argument rather than doing this?

mballance replied:
Sure, that's fine. We get a few warnings without it set, but that's okay

andrewvaughanj replied:
Do you mean Cython-level warnings? Like “blah blah blah language level not set”?

If so, that probably means Boolector’s build system isn’t passing in `--3` when calling Cython (and it should if we’re building for Py3).

(sorry for the vagueness; not at a computer)

andrewvaughanj commented:
> > My Lingeling PR isn’t merged; how are you using the commit hash from that?
> > Could we do something here like we do with the Windows patches to make this self-contained within the Boolector repo?
> 
> Apologies for the confusion, @andrewvaughanj. I wanted to point @mpreiner to a description of the issue resolved by moving to the latest lingeling code. It appears the issue has been resolved independent of your PR. The clone variable is now named 'cloned'.

Ah! All good :)

mpreiner commented on pypi/setup.py:
```diff
@@ -0,0 +1,306 @@
+
+from distutils import log
+from distutils.core import setup, Command
+#from Cython.Build import cythonize
```
Remove?

mballance commented on pypi/setup.py:
```diff
@@ -0,0 +1,306 @@
+#include "boolector/boolector.h"
+
+int main(void) {
+
```
Makes sense. I've added this and confirmed that the build still runs properly.

mpreiner commented on pypi/build.sh:
```diff
@@ -0,0 +1,76 @@
+
+cd pyboolector
+
+for py in cp35-cp35m cp36-cp36m cp37-cp37m cp38-cp38 cp39-cp39; do
+  echo "Python: ${py}"
+  python=/opt/python/${py}/bin/python
+  cd ${BUILD_DIR}/pyboolector
+  rm -rf src
+  cp -r ${BUILD_DIR}/boolector/src/api/python src
+  sed -i -e 's/override//g' \
+     -e 's/noexcept/_GLIBCXX_USE_NOEXCEPT/g' \
+     -e 's/\(BoolectorException (const.*\)/\1\n    virtual ~BoolectorException() _GLIBCXX_USE_NOEXCEPT {}/' \
+       src/pyboolector_abort.cpp
+  mkdir -p src/utils
+  cp ${BUILD_DIR}/boolector/src/*.h src
+  cp ${BUILD_DIR}/boolector/src/utils/*.h src/utils
+  $python ./src/mkenums.py ./src/btortypes.h ./src/pyboolector_enums.pxd
+  $python setup.py sdist bdist_wheel
+done
```
Just out of curiosity: Why can't the library produced by `./configure --python` be used here?

mballance replied:
The only way I could find to produce a 'wheel' (Python distribution package) that PyPi would accept was to use the Python build commands via setup.py. I think part of this is due to the package auditing that exists to ensure that packages posted on PyPi will truly be compliant with the platform specifications they purport to be compatible with.

mballance replied:
BTW, one option going forward (thinking Bitwuzla) would be to build all Python packages via a setup.py. The CMakeLists.txt would be slightly different, and would invoke Python on the setup.py instead of directly building the shared library. The advantage would be a single way of building the Python package.

mpreiner replied:
Oh ok. Do you have experience with scikit-build? Look like this is allows better integration with a cmake build system.

mballance replied:
Unfortunately, I don't. I think I looked at scikit-build early in my process of learning about building Python extensions, and think my conclusion was that it was optimized for certain types of extension but lacked flexibility. Certainly not in a position to offer a knowledgeable opinion, though.

I've been working on building Python extensions as part of the regular build in the context of a different project. Thus far, it doesn't look too difficult. Here's the CMakeList fragment that invokes the actual build:
https://github.com/PSSTools/pssparser/blob/67fdb8d0cd41874f80f6e8c1acb60a83ac6cc156/ext/CMakeLists.txt#L7-L20

This, in turn, is driven by a script run inside the 'manylinux2010' Docker image to produce all the Python-version variants:
https://github.com/PSSTools/pssparser/blob/67fdb8d0cd41874f80f6e8c1acb60a83ac6cc156/.github/workflows/pypi.sh#L26-L38

Fortunately, Makefile dependencies cause only the Python extension (and not all the other code) to be rebuilt for each Python version

andrewvaughanj commented on pypi/build.sh:
```diff
@@ -0,0 +1,76 @@
+  sed -i -e 's/override//g' \
+     -e 's/noexcept/_GLIBCXX_USE_NOEXCEPT/g' \
+     -e 's/\(BoolectorException (const.*\)/\1\n    virtual ~BoolectorException() _GLIBCXX_USE_NOEXCEPT {}/' \
+       src/pyboolector_abort.cpp
```
What's this bit for?

mballance replied:
In order to support a broad range of Linux distributions, pre-compiled Python libraries are built on specific older Linux distros with older compilers (Python defines a manylinux docker image based on CentOS). The supported compiler versions have limits on the C++ dialects supported (override, noexcept as a keyword, etc). This filter patches the code to comply with the compiler-supported C++.

mpreiner commented on pypi/azure-pipelines-pypi.yml:
```diff
@@ -0,0 +1,113 @@
+
+jobs:
```
Is there a way to only run the upload to pypi if we do a new release? I think PyPi has upload limits and we should not upload a new version on every commit.

mballance replied:
There should be, as long as we can detect creation of a tag or something else that marks the creation of a release. I always find I have to do a bit of Googling and playing around to get this to work, but there typically is a way. How are releases marked for Boolector? A tag? A branch?
That said, I've not detected any upload limits with PyPi on any of the projects I'm using it with.

mpreiner replied:
Releases are tagged, but there is also the releases page: https://github.com/Boolector/boolector/releases

If Azure doesn't have this kind of functionality we could also think of switching the PyPi upload to GitHub actions, which supports running workflows only for specific actions such as creating a release.

mpreiner reviewed (approved):
@mballance is there anything else to do other than what you wrote [here](https://github.com/Boolector/boolector/pull/72#issuecomment-555300534) once the PR is merged?

mballance commented:
> @mballance is there anything else to do other than what you wrote [here](https://github.com/Boolector/boolector/pull/72#issuecomment-555300534) once the PR is merged?

@mpreiner, other than configuring PyPi upload (referenced link), we should be good to go. Please let me know the PyPi username that will be doing uploads, since I think I'll need to add it to the PyPi project.
